[biomed nl] detection

chejennifer opened the pull request:
This PR has the detection changes for biomed NL. SV detector detects sv or prop & Place detector detects place or entity

Frontend changes to display these backend changes: https://github.com/datacommonsorg/website/pull/3999

https://github.com/datacommonsorg/website/assets/69875368/624e081d-4ec4-47ac-be22-2011cae79665

pradh reviewed (commented):
Any chance of breaking this up into backend vs. frontend, or further smaller blocks?

pradh commented on server/integration_tests/test_data/demo2_cities_feb2023/query_1/chart_config.json:
```diff
@@ -93,6 +93,7 @@
   },
   "context": {},
   "debug": {},
+  "entityPvConfig": null,
```
nit:  tripleConfig?

pradh commented on server/lib/nl/fulfillment/types.py:
```diff
@@ -159,3 +159,27 @@ class ChartSpec:
+
+@dataclass
+class EntityPvConfig:
+  entities: List[Place]
```
It need not be just `Place` right?

pradh commented on server/lib/nl/fulfillment/types.py:
```diff
@@ -159,3 +159,27 @@ class ChartSpec:
+  dcid: str
+  name: str
+  value: str
+  provenanceId: str
```
Use snake-case?

pradh commented on server/lib/nl/fulfillment/types.py:
```diff
@@ -159,3 +159,27 @@ class ChartSpec:
+
+
+@dataclass
+class PropertyValue:
```
Add comments?  Unclear if this is a triple or just values?  If the former, call it Triple?

pradh commented on server/lib/nl/fulfillment/types.py:
```diff
@@ -159,3 +159,27 @@ class ChartSpec:
+  entities: List[Place]
+  properties: List[Property]
+  # dict of entity dcid -> property id -> values for that property and entity
+  propertyValues: Dict[str, Dict[str, PropertyValue]]
```
snake

pradh commented on server/lib/nl/fulfillment/fulfiller.py:
```diff
@@ -95,6 +96,8 @@ def fulfill(uttr: Utterance, explore_mode: bool = False) -> PopulateState:
     elif main_qt == QueryType.COMPARISON_ACROSS_PLACES:
       # There are multiple places so we don't fallback.
       state.disable_fallback = True
+    elif main_qt == QueryType.PROPVALS:
```
`TRIPLE`?

pradh commented on server/lib/nl/detection/context.py:
```diff
@@ -55,7 +55,7 @@ def merge_with_context(uttr: nl_uttr.Utterance, default_place: Place = None):
   # TODO: Confirm type for ranking
   place_type = utils.get_contained_in_type(uttr)
 
```
Do we need context handling for entities?  Add TODO?

pradh commented on server/lib/nl/detection/place.py:
```diff
@@ -275,6 +276,16 @@ def get_place_from_dcids(place_dcids: List[str], debug_logs: Dict) -> any:
                              place_type=ptype,
                              country=country)
 
+  # If there were detected places that there wasn't any place info for, assume
```
Should we return entities separate from places?

Instead of `Place` type, it can be an `Entity` type?

pradh commented on server/lib/nl/detection/types.py:
```diff
@@ -31,6 +31,9 @@ class Place:
   name: str
   place_type: str
   country: str = None
+  # This is a temporary field to use as we test out treating other entities as
```
Should we have a separate type Entity?

pradh commented on server/lib/nl/detection/utils.py:
```diff
@@ -182,13 +182,22 @@ def create_utterance(query_detection: Detection,
   if (query_detection.places_detected) and (
       query_detection.places_detected.places_found):
-    uttr.places.extend(query_detection.places_detected.places_found)
+    places = []
```
Might be better for this to happen in detection?

pradh commented on server/lib/nl/fulfillment/fulfiller.py:
```diff
@@ -161,7 +164,12 @@ def _perform_strict_mode_checks(uttr: Utterance) -> bool:
+  query_types = []
+  if params.is_bio(uttr.insight_ctx) and uttr.entities:
+    query_types.append(QueryType.PROPVALS)
+    if not uttr.places:
```
We fall-through if there are places?  Do we then support fallback from PROPVALS to another entity type?

pradh commented on server/lib/nl/fulfillment/fulfiller.py:
```diff
@@ -95,6 +96,8 @@ def fulfill(uttr: Utterance, explore_mode: bool = False) -> PopulateState:
       # There are multiple places so we don't fallback.
       state.disable_fallback = True
+    elif main_qt == QueryType.PROPVALS:
+      done = propvals.populate(uttr)
```
I'm not 100% sure about this, explain in comments your rationale?

pradh commented on server/lib/nl/fulfillment/propvals.py:
```diff
@@ -0,0 +1,64 @@
+# Handler for Entity propval queries.
+#
+
+_SINGLE_ENTITY_SINGLE_PV_MULTI_VAL_TITLE = 'The {property}s for {entity} are:'
```
Since this is local for readability use shorter names?

pradh commented on server/lib/nl/fulfillment/propvals.py:
```diff
@@ -0,0 +1,64 @@
+
+def populate(uttr: nl_uttr.Utterance) -> bool:
+  if not uttr.entities:
+    uttr.counters.err('propvals_failed_noentities', 1)
```
Should we do more checks?  Is the detected thing property?

pradh commented on server/lib/nl/fulfillment/propvals.py:
```diff
@@ -0,0 +1,64 @@
+                        types=val.get('types', []))
+          for val in entity_triples[sv]
+      ]
+      title = _get_single_entity_single_pv_title(sv, entity.name or entity.dcid,
```
pass entity or use a variable?

pradh commented on server/lib/nl/fulfillment/fulfiller.py:
```diff
@@ -161,7 +164,12 @@ def _perform_strict_mode_checks(uttr: Utterance) -> bool:
 def _produce_query_types(uttr: Utterance) -> List[QueryType]:
-  query_types = [handlers.first_query_type(uttr)]
+  query_types = []
+  if params.is_bio(uttr.insight_ctx) and uttr.entities:
```
Where are we detecting property?  I'm confused.

pradh reviewed (commented):
First pass..

pradh commented on server/lib/nl/detection/place.py:
```diff
@@ -275,6 +276,16 @@ def get_place_from_dcids(place_dcids: List[str], debug_logs: Dict) -> any:
 
+  # If there were detected places that there wasn't any place info for, assume
+  # they are non-place entities and get info through property values.
+  missing_place_info = [p for p in place_dcids if not dcid2place.get(p)]
```
Should the determination of place vs. entity be based on the returned type?  (available in Mixer csv)

chejennifer commented:
Thanks for the first pass review. As discussed offline, I've moved differentiation of detected prop vs sv to the detection part of the code & I've updated this PR to only have the detection changes and will have a separate PR for fulfillment once we decide on the response shape.

pradh commented on server/integration_tests/test_data/detection_api_bio/debug_info.json:
```diff
@@ -0,0 +1,49 @@
+    ]
+  },
+  "props_matching": {
+    "SV": [
```
Weird that its `SV`, maybe have TODO?

pradh commented on server/lib/nl/common/debug_utils.py:
```diff
@@ -42,6 +42,12 @@ def result_with_debug_info(data_dict: Dict, status: str,
+  props_dict = {
+      'SV': query_detection.svs_detected.non_sv.svs,
+      'CosineScore': query_detection.svs_detected.non_sv.scores,
+      'SV_to_Sentences': query_detection.svs_detected.non_sv.sv2sentences
```
nit: `s/SV/prop` ?

chejennifer replied:
done

pradh commented on server/lib/nl/detection/detector.py:
```diff
@@ -213,23 +214,25 @@ def construct_for_explore(entities: List[str], vars: List[str], child_type: str,
                                      single_sv=dutils.VarCandidates(
                                          svs=vars,
                                          scores=[0.51] * len(vars),
                                          sv2sentences={}),
```
s/non_sv/prop/ ?

chejennifer replied:
done

pradh commented on server/lib/nl/detection/types.py:
```diff
@@ -56,6 +65,8 @@ class SVDetection:
   # Single SV detection.
   single_sv: dvars.VarCandidates
+  # Detected variables that are not Statistical Variable or Topic.
+  non_sv: dvars.VarCandidates
```
prop

pradh commented on server/lib/nl/detection/place.py:
```diff
@@ -299,6 +312,23 @@ def get_place_from_dcids(place_dcids: List[str], debug_logs: Dict) -> any:
+  if non_place_entities:
+    names = property_values(non_place_entities, 'name')
+    for e in non_place_entities:
+      eNames = names.get(e, [])
```
nit: skip camel case for variable names in py?

chejennifer replied:
done

pradh commented on server/lib/nl/detection/place.py:
```diff
@@ -299,6 +312,23 @@ def get_place_from_dcids(place_dcids: List[str], debug_logs: Dict) -> any:
+    for e in non_place_entities:
+      eNames = names.get(e, [])
+      eName = eNames[0] if len(eNames) > 0 else e
+      entities.append(Entity(dcid=e, name=eName, type=''))
```
Why is there no type, can we get the type from the API so we can make downstream decisions?

Add TODO?

chejennifer replied:
since we currently weren't using the type, I wanted to avoid the extra API call, but added TODO

pradh replied:
Hmmm, I thought it doesn't need an extra API call --  the `property_values` I think doesn't expose the type, but the API call returns it? https://api.datacommons.org/v2/node?key=AIzaSyCTI4Xz-UW_G2Q2RfknhcfdAnTHq5X5XuI&nodes=geoId/06&property=-%3EcontainedInPlace

chejennifer replied:
but that's just for the containedInPlace property, name doesn't give the type: https://api.datacommons.org/v2/node?key=AIzaSyCTI4Xz-UW_G2Q2RfknhcfdAnTHq5X5XuI&nodes=geoId/06&property=-%3Ename

pradh replied:
Yes, `name` won't because its an `object_value`.   `object_id` things should return type, even beyond `containedInPlace`.  For example: https://api.datacommons.org/v2/node?key=AIzaSyCTI4Xz-UW_G2Q2RfknhcfdAnTHq5X5XuI&nodes=geoId/06&property=-%3EadministrativeCapital

pradh replied:
And we only want types for such IDs.

chejennifer replied:
I'm confused, we have a list of ids that we want to get types for, why would it not be an extra API call?

pradh replied:
Ohh I see.  I guess then we perhaps want a single Triple out call. But for later PR...

pradh commented on server/lib/nl/detection/types.py:
```diff
@@ -33,6 +33,14 @@ class Place:
+  """Entity attributes."""
+  dcid: str
+  name: str
+  type: str
```
clarify in comment that it a non-place entity only?

chejennifer replied:
done

pradh commented on server/routes/nl/helpers.py:
```diff
@@ -208,6 +208,10 @@ def parse_query_and_detect(request: Dict, backend: str, client: str,
     return None, err_json
   counters.timeit('query_detection', start)
 
+  allow_non_place = False
```
nit: allow_entity, so that further down it reads better (`if not allow_non_place:`)

chejennifer replied:
done

pradh commented on server/routes/nl/helpers.py:
```diff
@@ -327,7 +336,7 @@ def prepare_response(utterance: nl_utterance.Utterance,
   data_dict = {
-      'place': ret_places_dict[0],
+      'place': ret_places_dict[0] if len(ret_places) > 0 else {},
       'places': ret_places_dict,
```
do we need 'entities' here?

pradh replied:
then, later is fine!

pradh commented on server/routes/nl/helpers.py:
```diff
@@ -365,6 +374,7 @@ def prepare_response_common(data_dict: Dict,
                             client: str = '') -> Dict:
   data_dict = dbg.result_with_debug_info(data_dict, status_str, detection,
                                          dbg_counters, debug_logs)
+  print(data_dict)
```
remove

chejennifer replied:
done

pradh commented on server/lib/nl/detection/utils.py:
```diff
@@ -133,18 +134,52 @@ def empty_svs_score_dict():
+# Takes the detected svs and returns
+# 1. sv candidates: svs that are Statistical Variable or Topic
+# 2. non sv candidates: any other detected svs.
+def _get_sv_and_non_sv_candidates(
```
s/non_sv/prop/?

chejennifer replied:
done

pradh commented on server/lib/nl/detection/utils.py:
```diff
@@ -133,18 +134,52 @@ def empty_svs_score_dict():
+    svs_scores_dict: Dict) -> tuple[dvars.VarCandidates, dvars.VarCandidates]:
+  sv_candidates = empty_var_candidates()
+  non_sv_candidates = empty_var_candidates()
+  sv_types = property_values(svs_scores_dict['SV'], 'typeOf')
```
Hmmm, given this, can we do this only when it is `allow_entities` ?

Generally, its nice to not do any extra API calls by default unless `bio_ft`

chejennifer replied:
done

pradh commented on server/lib/nl/detection/utils.py:
```diff
@@ -133,18 +134,52 @@ def empty_svs_score_dict():
+    # Topic.
+    is_sv = False
+    # We have some curated topics that are not a node in the kg, so assume topic
+    # if the sv starts with dc/topic
```
Or check for type `Property` and classify it as non_sv?  So we don't add one more dep on `dc/topic/'

chejennifer replied:
I chose to check for SV or Topic instead of checking for Property because I want to try using "->" for linked properties so those would not return me a type ..

pradh replied:
hmm, ok, that feels brittle.  Have a comment?

chejennifer replied:
done

pradh commented on server/lib/nl/fulfillment/propvals.py:
```diff
@@ -0,0 +1,64 @@
+# Copyright 2024 Google LLC
```
There are some pending comments in this file?

Should we rename it to triple?

pradh reviewed (commented):
Thanks for making the change!  This is much nicer!

pradh commented on server/lib/nl/detection/utils.py:
```diff
@@ -133,18 +134,63 @@ def empty_svs_score_dict():
+  sv_types = property_values(svs_scores_dict['SV'], 'typeOf')
+  for i, sv in enumerate(svs_scores_dict['SV']):
+    sv_type_list = sv_types.get(sv, [])
+    # a sv is considered a sv if any of its types are Statistical Variable or
```
nit: "An `sv` is considered an `sv` ..."

pradh reviewed (approved):
Thanks for the changes!

chejennifer commented:
thanks for the review!
